README.md
# SMLReaderVZ #

ESP8266 based SML reader to fetch data of (electrical) meters via the optical interface and transfer it to Volkszaehler Middleware.

# Description

## Overview
Receive SML data from an (electrical) meter, evaluate it and send selected data to the middleware
of Volkszaehler (addresses and channel UUIDs configurable).

Evaluation is done in periods defined in structure *SensorConfig SENSOR_CONFIGS*.  
A webserver dashboard is available to show the instantaneous information.  

The build in LED is used for signalling:  
- Setup()
  - On at begin
  - Off before start of conf.init and sensor.init
  - on at end
- Loop()  
  - no switch
- wifi_connected()  
  - Off at begin
- process_message (SML data)  
  - On at begin
  - Off at end

Serial monitor output is available as configured during the build.

## Hardware
The target hardware is an Expressif ESP8266 microprocessor with a simple IR reading head that is attached to the optical interface of an electrical meter.  

### Layout
The hardware layout follows [3], connection PINs are defined in *config.h*.  

<img src="./doc/img/SMLreaderLayout.png" alt="Layout"  style="height: 423px; width:650px;" />

## Dash Board
<img src="./doc/img/SMLReaderVZ-DashBoard.png" alt="Dash Board"/>

## Configuration

### Web Interface
The program provides a Web Interface for configuration and a small home page.

### Web Access Point (WiFI AP mode)
The program provides a Web Access Point at boot time via http://192.168.4.1/config (or the configured SSID name) using the configured password.  
Note: http://192.168.4.1 will access the Dash Board, not the config page.  
It offers a configuration page both for the Access Point and for a local WLAN SSID name and password.  
Additional customer parameters are supported.  
Configuration is stored in EEPROM.  

At first boot, the defined default password *MY_WIFI_AP_DEFAULT_PASSWORD* (as defined in *config.h*) is used for AP mode access.  
Note: at first boot you need to configure the device: set a new AP password, the WLAN SSID and WLAN password, and push the apply button.
**Disconnect** from the the Access Point to continue and **reset** the device.

If no client connects before the timeout (configured to 30sec), the device will automatically continue in 
STA (station) mode and connect to a local WLAN if configured.  

### Web Interface in a local WLAN (WiFi STA mode)
If the device has already been configured, it will automatically connect to the local WiFi after timeout.  
The device web interface can be reached via the IP address obtained from your local network's DHCP server or the configured SSID name.  
A small home page is provided at *\<localIP\>/start* which offers access to the configuration page as well.  
To login into configuration, provide the user *admin* and the configured AP *password*.  

### Configuration Parameter

The configuration page provides two sections:
- System Configuration: WiFi AP/STA names and passwords
- VZ Settings: volkszaehler server name (or IP), volkszaehler middleware (e.g. middleware.php), uuid of selected data channels (including a channel for test data and a hearbeat channel) and a timezone offset.  
You can switch-off transmission of data by using "null" as uuid (configurable by VZ_UUID_NO_SEND in config.h)  
Note: SMLReaderVZ will send data with standard UNIX epochtime (ms) timestamps (ignoring timezone offset).

<img src="./doc/img/configUI.png" alt="Layout"/>

## Usage
- Just switch on the board.
- Connect to the access point if you need to configure your local WiFi net:  
IP address 192.168.4.1/config (or the specified network name) using your configured AP password.
- A force to AP mode is provided in case of forgotten password:  
Connect *WEBCONF_AP_MODE_CONFIG_PIN* to ground and use *MY_WIFI_AP_DEFAULT_PASSWORD*.  
- Wait for AP timeout and access the home page in your local network:  
accessible by MY_WIFI_AP_SSID (or IP address provided by your DHCP server).
- You can access the configuration page in STA mode by login as *admin* with the configured AP password.

### Serial Monitor Output
The amount of debug output to the serial monitor can be controlled by two DEFINE statements, e.g. given as compiler build flags.  
- *SERIAL_DEBUG=true* provides increased debug output
- *SERIAL_DEBUG_VERBOSE=true* provides output of the complete SML message block sent by the meter device. This allows to check which data is provided by the meter.

## Dash Board
A dash board is acessible both from AP and STA mode.
See https://github.com/ayushsharma82/ESP-DASH.


## Implementation
Using classes  
**Sensor:**      receive data and put it into a buffer  
**SmlHttp:**     transfers data to Volkszaehler data base  
**smlDebug:**    functions for output of sml messages to serial monitor [3]  

Used own libs:  
**confWeb**             configurable web server (derived from [1])  
**libSML**              parse and evaluate SML messages (error correction of [4])  
**ESP-Dash**            Dash Board [5]  

Used 3rd party libs:  
**ArduinoJson**  
**ESP AsyncWebServer**  provide web server on D1 mini  
**ESPAsyncTCP**  
**TimeLib**             date/time functions  

## Dependency Graph V2.0.0 ##
```bash
Dependency Graph
|-- ESP-DASH @ 4.0.1+sha.9431138
|   |-- ESP Async WebServer @ 1.2.3+sha.f71e3d4
|   |   |-- ESPAsyncTCP @ 1.2.2+sha.1547686
|   |   |-- Hash @ 1.0
|   |   |-- ESP8266WiFi @ 1.0
|   |-- ArduinoJson @ 6.20.1+sha.33a4773
|   |-- ESP8266WiFi @ 1.0
|   |-- ESPAsyncTCP @ 1.2.2+sha.1547686
|   |-- Hash @ 1.0
|-- ESPAsyncTCP @ 1.2.2+sha.1547686
|-- ESP Async WebServer @ 1.2.3+sha.f71e3d4
|   |-- ESPAsyncTCP @ 1.2.2+sha.1547686
|   |-- Hash @ 1.0
|   |-- ESP8266WiFi @ 1.0
|-- ArduinoJson @ 6.20.1+sha.33a4773
|-- Time @ 1.6.1+sha.a18e50d
|-- libSML @ 1.0.0+sha.74e18fc
|-- EspSoftwareSerial @ 7.0.0
|-- DNSServer @ 1.1.1
|   |-- ESP8266WiFi @ 1.0
|-- ESP8266WiFi @ 1.0
|-- EEPROM @ 1.0
|-- Hash @ 1.0
|-- confWeb @ 1.0.0
|   |-- ESP Async WebServer @ 1.2.3+sha.f71e3d4
|   |   |-- ESPAsyncTCP @ 1.2.2+sha.1547686
|   |   |-- Hash @ 1.0
|   |   |-- ESP8266WiFi @ 1.0
|   |-- ArduinoJson @ 6.20.1+sha.33a4773
|   |-- EEPROM @ 1.0
|   |-- ESP8266WiFi @ 1.0
|   |-- ESPAsyncTCP @ 1.2.2+sha.1547686
|   |-- Hash @ 1.0
|   |-- DNSServer @ 1.1.1
|   |   |-- ESP8266WiFi @ 1.0
|-- ESP8266HTTPClient @ 1.2
|   |-- ESP8266WiFi @ 1.0
```

# Description Class Sensor #
Class Sensor initializes the serial input PIN and transfers the serial input into a buffer.

## Usage ##
in Setup():
Create a Sensor object, providing the configuration structure and a call back function to process the message  

	Sensor \*sensor = new Sensor(config, process_message);

in Loop():  

	sensor->loop()

## Implementation ##
A state machine is used to
- wait for incoming data by checking for the SML start sequence
- transfer data to the buffer until the end sequence is recognized
- handle the CRC data
- initiate processing of the data using the callback function.

## Used libs ##
SoftwareSerial  

## Details ##
### State machine of Sensor.cpp ###

| Old State            | Event                                            | New State                     |
|:---------------------|:------------------------------------------------:|------------------------------:|
|                      | declaration                                      | 0 INIT                        |  
| 0 INIT               | init_state()                                     | 2 WAIT_FOR_START_SEQ          |  
| 2 WAIT_FOR_START_SEQ | check Read Timeout: no timeout                   | 2 WAIT_FOR_START_SEQ          |  
| 2 WAIT_FOR_START_SEQ | check Read Timeout: timeout, reset_state()       | 0 INIT                        |  
| 2 WAIT_FOR_START_SEQ | start sequence found                             | 3 READ_MESSAGE                |  
| 3 READ_MESSAGE       | END SEQUENCE found                               | 5 READ_CHECKSUM               |  
| 3 READ_MESSAGE       | buffer overflow, reset_state()                   | 0 INIT                        |  
| 5 READ_CHECKSUM      | read checksum                                    | 4 PROCESS_MESSAGE (callback)  |  
| 4 PROCESS_MESSAGE    | pause interval == 0, reset_state()               | 0 INIT                        |  
| 4 PROCESS_MESSAGE    | pause interval != 0, reset_state()               | 1 STANDBY                     |  
| 1 STANDBY            | standby time over, reset_state()                 | 0 INIT                        |  


Remark: reset_state() calls init_state() which results in 2 WAIT_FOR_START_SEQ; state INIT is only used during initialisation;

# Description of class SmlHttp #
Perform http transfer of a data tupel (timestamp,value) of a sensor channel to the Volkszaehler data base via middleware.php  
- timestamp is Unix epoch time in seconds.  
- sensor channel is defined by its data base UUID.  

This class replaces class MqttPublisher that was used in https://github.com/mruettgers/SMLReader  as a http server is used instead of an MQTT broker.  
The http implementation follows the example in https://RandomNerdTutorials.com/esp8266-nodemcu-http-get-post-arduino/.

## Usage ##
```bash
myHttp.init(SmlHttpConfig &config)              // initialize class with server name and channel UUIDs
myHttp.postHttp(vzUUID, s_timeStamp, value);    // post value to Volkszaehler
myHttp.publish(sensor, file);                   // evaluate and filter SML file messages and call postHttp()
myHttp.testHttp();                              // create test output and call postHttp()
myHttp.getTimeStamp();                          // returns TimeStamp string
myHttp.getValue(obisValue _select);             // returns selected Obis value of an SML message, valid only with publish()
```
Server name and Volkszaehler channel UUIDs are provided via struct SmlHttpConfig.


## Implementation ##

postHttp():  
For transfer to volkszaehler, the http transfer should look like this:
```bash
http://volkszaehler-server/middleware.php/data.json?uuid=ae53c580-1234-5678-90ab-cdefghijklmn&operation=add&ts=1666801000000&value=22
```

Implementation is done using class HTTPClient.

publish():  
The publish() method evaluates the SML messages of the SML file structure extracting Obis name of channels and the data.  
The timestamp is created locally based on the system time.  
Sensor is only used to extract configuration data (name of meter, numeric flag).

# Description smlDebug.cpp #
smlDebug.cpp implementing debug output of the SML file message structure.

# License
Copyright (C) M. Herbert, 2022-2023.  
Licensed under the GNU General Public License v3.0

# Acknowledgement

This project is based on https://github.com/mruettgers/SMLReader.  
It was adapted for own use with these changes:  
- modularization of code (own cpp instead of monolithic include of code into main.cpp)
- removal of jled usage
- use ESP-Dash (with ESPAsyncWebserver instead of IOTwebserver)
- remove MQTT interface and add http interface to Volkszaehler middleware instead  
- added confWeb for the configuration interface in WiFi AP mode  

# History
see [CHANGELOG.md](http://./CHANGELOG.md)

# References
[1] https://github.com/prampec/IotWebConf  
[2] https://github.com/10k-resistor/Solis4Gmini-logger  
[3] https://github.com/mruettgers/SMLReader  
[4] https://github.com/volkszaehler/libsml  
[5] https://github.com/ayushsharma82/ESP-DASH  
[6] https://RandomNerdTutorials.com/esp8266-nodemcu-http-get-post-arduino/



CHANGELOG.md
# Changelog #
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Released] ##

## [2.2.1] - 2023-02-19 ##
### Fixed
- add missing update of date/time in loop


## [2.2.0] - 2023-02-17 ##
### Changed ###
- split VZ_SERVER and VZ_MIDDLEWARE names, use pure UUID strings
- config UI: separate config of server and middleware; add TIMEZONE offset to support switch to daylight saving time
- no transmission for data with UUID = VZ_UUID_NO_SEND
- remove use of timeControl and NTPclient (use built-in timer function of Arduino)
- adapt home page
- use original ESP-Dash

## [2.1.0] - 2023-01-29 ##
### Changed ###
- use timeControl 1.1.0 (built-in ntp timer, error corrections)
- add heartbeat channel
- rename obisValue to UuidValueName in smlHttp
- add Title card on Dash Board
- remove webConfMode in main()

### Fixed ###
-

## [2.0.0] - 2023-01-29 ##
### Changed ###
- added confWeb for the configuration interface in WiFi AP mode
- added struct MyHttpConfig for the configuration of VZ server and UUID
- added Home Page and handler
- added in class MyHttp: _uuid\[\], method setServerName(), testHttp(), initHttp()
- rename MyHttp.cpp/.h to smlHttp.cpp/.h and class to SmlHttp because strong dependance on smlLib.
- use lib timeControl instead of local time.cpp, update to new interface
- use own (modified) ESP-Dash @ 4.0.1
- clean up cascaded include files
- removed MicroDebug lib, defined Macros in respective include files.
- remove usage of toggleBuiltinLED

### Fixed ###
- name collision of hexdump in libSML/sml_shared.c/.h: renamed to sml_hexdump;  
allows to use newest Arduino framework
- Sensor.cpp: disables namespace std; added std:: to unique_ptr<SoftwareSerial>
  reason: byte was ambiguous
- rename local debug.c/.h to smlDebug.c/.h due to collision with Arduino.h
- smlDebug.h: removed #include <FormattingSerialDebug.h>, defined Macros there
  reason: problem with #pragma statement.


## [1.0.0] - 2022-12-08 ##
### Changed ###
- first version
### Fixed ###
-



src/smlHttp.cpp
#include <string.h>
#include <ESP8266HTTPClient.h>
#include "config.h"
#include "smlHttp.h"
#include "smlDebug.h"

/* *** smlHttp.cpp

transfer data to and from a web server

2023-02-27 mh
- split up input for server url
- not transmission, if uuid = VZ_UUID_NO_SEND

2023-02-05 mh
- rename obisValue to UuidValueName

2023-01-26 mh
- added setServerName(), removed VZ_SERVER and VZ_UUID_*
- added init(SmlHttpConfig &config)
- added testHttp(): test channel to database
- rename from MyHttp.cpp/.h to smlHttp.cpp/.h and class to SmlHttp because strong dependance on smlLib.

2022-12-08 mh
- use double value for postHttp to avoid loose of accurancy of SML meter data

2022-11-30 mh
- added method publish to evaluate and transfer SML message data to Volkszaehler
- based on https://github.com/mruettgers/SMLReader changing the mqtt publish method

2022-11-09	M. Herbert
- based on https://randomnerdtutorials.com/esp8266-nodemcu-http-get-post-arduino/ (see Copyright notice below)
- modified for own use


Copyright (C) M. Herbert, 2022-2023.
Licensed under the GNU General Public License v3.0

*** end change log *** */

/* ***
# Description of class SmlHttp #
http transfer of a data tupel (timestamp,value) of a sensor channel to the Volkszaehler data base via middleware.php  
- timestamp is Unix epoch time in seconds.
- sensor channel is defined by its data base UUID.

This class replaces class MqttPublisher that was used in https://github.com/mruettgers/SMLReader  as a http server is used instead of an MQTT broker.  
The http implementation follows the example in https://RandomNerdTutorials.com/esp8266-nodemcu-http-get-post-arduino/N_UUID_VALUE.  

## Usage ##
```bash
myHttp.init(SmlHttpConfig &config)              // initialize class with server name and channel UUIDs
myHttp.postHttp(vzUUID, s_timeStamp, value);    // post value to Volkszaehler
myHttp.publish(sensor, file);                   // evaluate and filter SML file messages and call postHttp()
myHttp.testHttp();                              // create test output and call postHttp()
myHttp.getTimeStamp();                          // returns TimeStamp string
myHttp.getValue(UuidValueName _select);             // returns selected Obis value of an SML message, valid only with publish()
```
Server name and Volkszaehler channel UUIDs are provided via struct SmlHttpConfig.

## Implementation ##

postHttp():  
For transfer to volkszaehler, the http transfer should look like this:
```bash
// http://volks-raspi/middleware.php/data.json?uuid=ae53c580-5549-11ed-84a0-cfe6bdf4d646&operation=add&ts=1666801000000&value=22
```

Implementation is done using class HTTPClient.

publish():  
The publish() method evaluates the SML messages of the SML file structure extracting Obis name of channels and the data.  
The timestamp is created locally based on the system time.  
Sensor is only used to extract configuration data (name of meter, numeric flag).

*** end description *** */

/*
  License notice Rui Santos
  Complete project details at Complete project details at https://RandomNerdTutorials.com/esp8266-nodemcu-http-get-post-arduino/

  Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files.
  The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
  
  Code compatible with ESP8266 Boards Version 3.0.0 or above 
  (see in Tools > Boards > Boards Manager > ESP8266)
*/

String baseTopic="";      // was used as root for MQTT
WiFiClient _client;
HTTPClient http;
// const char* _serverName="http://volks-raspi/middleware.php/data.json";

SmlHttp::SmlHttp()
{
  uint16_t i;
  for (i=0;i<N_UUID_VALUE;i++)
  {
    _value[i]=0.;
  }
}

void SmlHttp::init(SmlHttpConfig &config) {
  _serverName = String(config.vzServer);
  _middlewareName = String(config.vzMiddleware);
  DEBUG_TRACE(VERBOSE_LEVEL_HTTP,"vzServer: %s",_serverName.c_str());
  DEBUG_TRACE(VERBOSE_LEVEL_HTTP,"vzMiddleware: %s",_middlewareName.c_str());

  uint16_t i;
  for (i=0;i<N_UUID_VALUE;i++)
  {
    _uuid[i] = &config.uuidValue[i][0];
    DEBUG_TRACE(VERBOSE_LEVEL_HTTP,"uuid[%d] = %s",i,_uuid[i]);
  }

};
void SmlHttp::setServerName(String serverName) {
  _serverName = serverName;
};

void SmlHttp::setMiddlewareName(String middlewareName)
{
  _middlewareName = middlewareName;
};

int SmlHttp::postHttp(String vzUUID, String timeStamp, double value)
{
    //For transfer to volkszaehler, the http transfer should look like this:
    // http://volks-raspi/middleware.php/data.json?uuid=ae53c580-1234-5678-90ab-cdefghijklmn&operation=add&ts=1666801000000&value=22

  if(!strcmp(vzUUID.c_str(),VZ_UUID_NO_SEND))
  {
    return -99;
  }
  String vzUrl = "http://";
  vzUrl += _serverName + "/" + _middlewareName;
  vzUrl += "/" + String(VZ_DATA_JSON);

  if(!http.begin(_client, vzUrl))
  {
    DEBUG_TRACE(VERBOSE_LEVEL_HTTP,"No connection to %s",_serverName.c_str());
    return 404;
  };

  //http.setAuthorization("REPLACE_WITH_SERVER_USERNAME", "REPLACE_WITH_SERVER_PASSWORD");

  this->_TimeStamp = timeStamp + "000";    // store internally in ms

  // HTTP request with a content type: x-www-form-urlencoded
  http.addHeader("Content-Type", "application/x-www-form-urlencoded");

  //construct the message body
  //example for data to be sent: uuid=ae53c580-1234-5678-90ab-cdefghijklmn&operation=add&ts=1666801000000&value=22

   String httpRequestData = "uuid=" + vzUUID;
   String httpOperation = "&operation=add";
   
   String s_timestamp = "&ts=";
   s_timestamp += timeStamp;
   s_timestamp += "000";              // convert seconds to milli seconds
   
   String s_value = "&value="; 
   s_value += String(value);
   
   httpRequestData = httpRequestData + s_timestamp + s_value;

  DEBUG_TRACE(VERBOSE_LEVEL_HTTP,"Post message: %s",httpRequestData.c_str());

  int httpResponseCode = http.POST(httpRequestData);

  DEBUG_TRACE(VERBOSE_LEVEL_HTTP,"HTTP Response code: %d",httpResponseCode);
      
  // Free resources
  http.end();
  return httpResponseCode;
};

void SmlHttp::publish(Sensor *sensor, sml_file *file)
{

    for (int i = 0; i < file->messages_len; i++)
    {
      sml_message *message = file->messages[i];
      if (*message->message_body->tag == SML_MESSAGE_GET_LIST_RESPONSE)
      {
        sml_list *entry;
        sml_get_list_response *body;
        body = (sml_get_list_response *)message->message_body->data;
        for (entry = body->val_list; entry != NULL; entry = entry->next)
        {
          if (!entry->value)
          { // do not crash on null value
            continue;
          }

          char obisIdentifier[32];
          char buffer[255];
          String s_timestamp;
          String _vzUUID;

          sprintf(obisIdentifier, "%d-%d:%d.%d.%d*%d",              // adapted to VZ, original was: "%d-%d:%d.%d.%d/%d"
                  entry->obj_name->str[0], entry->obj_name->str[1],
                  entry->obj_name->str[2], entry->obj_name->str[3],
                  entry->obj_name->str[4], entry->obj_name->str[5]);

          // construction of MQTT path - currently used only for DEBUG
          String entryTopic = baseTopic + "sensor/" + (sensor->config->name) + "/obis/" + obisIdentifier + "/";

          // check for time stamp or use local time
          // Note: my meter does not send time, therefore we use local time
#if 0
          struct timeval tv;      // defined in time.h
          if (entry->val_time) { /* use time from meter */
            tv.tv_sec = *entry->val_time->data.timestamp;
            tv.tv_usec = 0;
            DEBUG("Use meter time %ldsec %ldus",tv.tv_sec,tv.tv_usec);
          } else {
            gettimeofday(&tv, NULL); /* use local time */
            DEBUG("Use local time %ldsec %ldus",tv.tv_sec,tv.tv_usec);
          }
#endif
          struct timeval tv;                      // defined in time.h
          gettimeofday(&tv, NULL);                // use local time; note that usec also contains ms --> divide by 1000 to get ms
          s_timestamp = String(tv.tv_sec);        // timestamp with resolution of 1 sec

          if (((entry->value->type & SML_TYPE_FIELD) == SML_TYPE_INTEGER) ||
              ((entry->value->type & SML_TYPE_FIELD) == SML_TYPE_UNSIGNED))
          {
            double value = sml_value_to_double(entry->value);
            int scaler = (entry->scaler) ? *entry->scaler : 0;
            int prec = -scaler;
            if (prec < 0)
              prec = 0;
            value = value * pow(10, scaler);
            sprintf(buffer, "%.*f", prec, value);
            DEBUG("%s: %s",entryTopic.c_str(), buffer);   // buffer contains the value as string in float format

//            publish(entryTopic + "value", buffer);   /* old, for MQTT */

            // we publish only numeric data, other parts below are kept for future use
            // we are interested only in specific data

            if( 0 == strcmp(obisIdentifier,OBIS_ID_ENERGY_IN))
            {
              _vzUUID = String(_uuid[vzENERGY_IN]);
              this->postHttp(_vzUUID, s_timestamp, value);
              // this->_TimeStamp = s_timestamp; done in postHttp()
              this->_value[vzENERGY_IN] = value;
            }
            else if( 0 == strcmp(obisIdentifier,OBIS_ID_ENERGY_OUT))
            {
              _vzUUID = String(_uuid[vzENERGY_OUT]);
              this->postHttp(_vzUUID, s_timestamp, value);
              // this->_TimeStamp = s_timestamp;  done in postHttp()
              this->_value[vzENERGY_OUT] = value;
            }
            else if( 0 == strcmp(obisIdentifier,OBIS_ID_POWER_IN))
            {
              _vzUUID = String(_uuid[vzPOWER_IN]);
              this->postHttp(_vzUUID, s_timestamp, value);
              // this->_TimeStamp = s_timestamp;  // done in postHttp()
              this->_value[vzPOWER_IN] = value;
            }
            else
            {
              /* do nothing */
            }
          }
          else if (!sensor->config->numeric_only)
          {
            if (entry->value->type == SML_TYPE_OCTET_STRING)
            {
              char *value;
              sml_value_to_strhex(entry->value, &value, true);
//             publish(entryTopic + "value", value);
              DEBUG("%s: %s",entryTopic.c_str(), value);

              free(value);
            }
            else if (entry->value->type == SML_TYPE_BOOLEAN)
            {
//              publish(entryTopic + "value", entry->value->data.boolean ? "true" : "false");
              DEBUG("%s: %s",entryTopic.c_str(), entry->value->data.boolean ? "true" : "false");
            }
          }
        }
      }
    }
}

String SmlHttp::getTimeStamp()
{
  return _TimeStamp;
}
double SmlHttp::getValue(UuidValueName _select)
{
  return _value[_select];
}
void SmlHttp::testHttp()
//
// 2023-01-26 mh
// - test channel to database, use system time as test data
{
  static uint32_t lastSendTime=0;
  uint32_t currentTime;
  currentTime = millis();
  if((currentTime-lastSendTime) > MY_TEST_SEND_UPDATE)
  {
    lastSendTime = currentTime;
    String _vzUUID = String(_uuid[vzTEST]);
      struct timeval tv;                      // defined in time.h
      gettimeofday(&tv, NULL);                // use local time; note that usec also contains ms --> divide by 1000 to get ms
      String s_timestamp = String(tv.tv_sec);        // timestamp with resolution of 1 sec

    this->postHttp(_vzUUID, s_timestamp, double(currentTime/1000.));
    // this->_TimeStamp = s_timestamp;  // done in postHttp()
    this->_value[vzTEST] = double(currentTime/1000.);
  }
}
